### PR TITLE
Extend the Type Cast code action for unsupported binary operations

### LINKED ISSUE #33912: Extend type cast code action to support undefined numeric operations
**Description:**
Consider the following scenario
screen

Here the denominator `entry[1]` can be casted to `float`, yet we don't suggest the code action for it. 

```ballerina
function findTop10GoldMedalsToAthletesRatio() returns error? {
    map<int> athletesByCountry = check findNumOfAthletesBycountry();
    stream<MedalStats, error?> medalStats = check getMedalStats();

    var result = from [string,int] entry in athletesByCountry.entries() 
    join var stats in medalStats on entry[0] equals stats.country 
    let float ratio = (<float>stats.totalMedals / entry[1]) * 100 
    select  
}

```

### PR DESCRIPTION
**Pull request by malinthar**
## Purpose
$subject to add a new code action to add type cast expr for unsupported binary operations.

Fixes #33912

## Samples
Peek 2022-05-12 00-05

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [x] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionContextTypeResolver.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package org.ballerinalang.langserver.codeaction;
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionContextTypeResolver.java:1]: Missing license header

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/TypeCastNumericExpression.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
+package org.ballerinalang.langserver.codeaction.providers.changetype;
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/TypeCastNumericExpression.java:1]: Missing license header

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/TypeCastNumericExpression.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        Optional<ExpressionNode> expressionNode = expressionResolver.findExpression(positionDetails.matchedNode());
+        if (operatorKind.isEmpty() || lhsOperand.isEmpty() || rhsOperand.isEmpty()
+                || expressionNode.isEmpty() || expressionNode.get().kind() != SyntaxKind.BINARY_EXPRESSION) {
+            return Collections.emptyList();
+        }
+        BinaryExpressionNode binaryExpressionNode = (BinaryExpressionNode) expressionNode.get();
+
+        //Check if the binary expression is used for an assignment. 
+        // If so, check if the type of the variable and the types of lsh and rhs operands is a 
+        // match and select te operand to cast accordingly.
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/TypeCastNumericExpression.java:65]:
```suggestion
        // match and select the operand to cast accordingly.
```

**Review by TharushiJay — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/TypeCastNumericExpression.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                new MatchedExpressionNodeResolver(positionDetails.matchedNode());
+        Optional<ExpressionNode> expressionNode = expressionResolver.findExpression(positionDetails.matchedNode());
+        if (operatorKind.isEmpty() || lhsOperand.isEmpty() || rhsOperand.isEmpty()
+                || expressionNode.isEmpty() || expressionNode.get().kind() != SyntaxKind.BINARY_EXPRESSION) {
+            return Collections.emptyList();
+        }
+        BinaryExpressionNode binaryExpressionNode = (BinaryExpressionNode) expressionNode.get();
+
+        //Check if the binary expression is used for an assignment. 
+        // If so, check if the type of the variable and the types of lsh and rhs operands is a
```
Comment (mindula) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/TypeCastNumericExpression.java:64]:
```suggestion
        // If so, check if the type of the variable and the types of lhs and rhs operands is a 
```

File: language-server/modules/langserver-core/src/test/resources/codeaction/type-cast/source/type_cast_in_binary_operation.bal [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    let decimal x1 = f1 / d1;
+
+function testTypeCastPositionalArg() {
+    int addResult = add(i1/f1,i1);
+}
+
+
+function add(float a, int b) returns int {
+    return <int> a + b;
+}
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-core/src/test/resources/codeaction/type-cast/source/type_cast_in_binary_operation.bal:25]: Shall we add a newline here?

**Review by gayalkuruppu — DISMISSED** [review]

**Review by IMS94 — DISMISSED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionContextTypeResolver.java [hunk 6]
```diff
@@ -34,6 +34,7 @@
 /**
  * This visitor is used to resolve the type of given code action context.
  * 
+ * Todo: Use the ContextTypeResolver instead.
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionContextTypeResolver.java:37]: @malinthar Did we create an issue for this?

Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionContextTypeResolver.java:37]: I will create an issue.

**Review by IMS94 — APPROVED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionContextTypeResolver.java (hunk 6 above)
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionContextTypeResolver.java:37]: Thanks!

**Pull request merged**